### PR TITLE
test: Added calcualtion logic for positve grid loss

### PR DESCRIPTION
**Pull request by energinet-ajw**
# Description

Added test positive_gris_loss - when period is one hour.
Added 3 test factories.

refac: Moved commen imports to utils file and added utils structure.

## Pull-request quality

- [x] The title adheres to [this guide](https://github.com/Mech0z/GitHubGuidelines)
- [x] Tests are written and executed locally
- [ ] Subsystem tests have been tested (by manually deploying to `dev_002` or `sandbox_002`)
- [ ] Documentation has been updated
- [ ] The integration [event catalog](https://energinet.atlassian.net/wiki/spaces/D3/pages/555581556/Event+catalog) has been updated
- [ ] C4 diagrams have been updated and Team Outlaws has been informed about relevant changes

### CODE REVIEW

File: source/databricks/calculation_engine/tests/calculation_logic/utils/factories/input_metering_point_periods_test_factory.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    BALANCE_RESPONSIBLE_ID = "1234567890123"
+
+    def __init__(self, spark: SparkSession):
+        self.spark = spark
+
+    @staticmethod
+    def create_row(
+        metering_point_id: str = METERING_POINT_ID,
+        metering_point_type: InputMeteringPointType = METERING_POINT_TYPE,
+        calculation_type: CalculationType | None = CALCULATION_TYPE,
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/calculation_logic/utils/factories/input_metering_point_periods_test_factory.py:51]: I believe this must always be None and should actually be removed from input tables. So perhaps you should simply remove it. Or at least hard code it to None.

File: source/databricks/calculation_engine/tests/calculation_logic/utils/factories/input_metering_point_periods_test_factory.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        settlement_method: InputSettlementMethod = SETTLEMENT_METHOD,
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/calculation_logic/utils/factories/input_metering_point_periods_test_factory.py:52]: Can be null

Comment (energinet-ajw) [reply source/databricks/calculation_engine/tests/calculation_logic/utils/factories/input_metering_point_periods_test_factory.py:52]: Done

File: source/databricks/calculation_engine/tests/calculation_logic/utils/factories/input_metering_point_periods_test_factory.py (hunk 1 above)
Comment (energinet-ajw) [reply source/databricks/calculation_engine/tests/calculation_logic/utils/factories/input_metering_point_periods_test_factory.py:51]: The default value was None. Now it is hardcoded to None.

File: source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/readme.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  AND one exchange metering point  
+  AND one production metering point
+WHEN calculating positive grid loss for 1st of February from 12pm to 1pm
+THEN the calculated positive grid loss is 3.750 kWh per quarter for the 1 hour period 
+THEN there are 4 rows in the result
+```
+
+```text
+The grid loss is calculated like this:
+(Sum E18 + (Sum E20 in - Sum E20 out) - (Sum E17 non-profiled + Sum E17 flex)) / 4 = grid loss
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/readme.md:15]: We should use "production" instead of E18 etc.

Comment (BjarkeMeier) [reply source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/readme.md:15]: How about using Σ?

File: source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/readme.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# Calculate positive grid loss
+
+```text
+GIVEN grid loss and system correction metering points (positive and negative grid loss)
+  AND one consumption metering point
+  AND one exchange metering point  
+  AND one production metering point
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/readme.md:7]:
```suggestion
  AND one production metering point
  AND one hourly time series point for each metering point
```

File: source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/readme.md (hunk 3 above)
Comment (energinet-ajw) [reply source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/readme.md:15]: Changed.

File: source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/states/scenario_state.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        quantity=10.0,
+        quality=QuantityQuality.MEASURED.value,
+        observation_time=datetime(2023, 2, 1, 12, 0, 0),
+    )
+    return factory.create_dataframe([row1, row2, row3])
+
+
+def create_metering_point_periods(spark: SparkSession) -> DataFrame:
+    factory = clf.InputMeteringPointPeriodsTestFactory(spark)
+    row1 = factory.create_row(
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/states/scenario_state.py:68]: It would be easier to read with meaningful names. E.g.
```suggestion
    consumption_period = factory.create_row(
```
or "consumption = ..."

Comment (BjarkeMeier) [reply source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/states/scenario_state.py:68]: Actually, I think this is way better - took me a while to figure out:
```suggestion
    positive_grid_loss = factory.create_row(
```

Comment (BjarkeMeier) [reply source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/states/scenario_state.py:68]: But this leads to another thought:
```suggestion
    positive_grid = factory.create_positive_grid_loss_row(...)
```
where the following input parameters are removed:
- metering_point_type, settlement_method

File: source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/states/scenario_state.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        metering_point_id="571313180480500149",
+        metering_point_type=InputMeteringPointType.PRODUCTION,
+        grid_area="804",
+        resolution=MeteringPointResolution.QUARTER,
+        energy_supplier="8100000000108",
+        balance_responsible_id="8100000000207",
+        from_date=datetime(2023, 1, 31, 23, 0, 0),
+    )
+
+    row3 = factory.create_row(
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/states/scenario_state.py]: How about
```suggestion
    row3 = factory.create_exchange_row(
```
and then remove the from_grid_area and to_grid_area from the regular create_row function?

File: source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/test_data/expected_results.csv [hunk 7]
```diff
@@ -0,0 +1,5 @@
+﻿grid_area;energy_supplier_id;balance_responsible_id;quantity;quantity_qualities;time;aggregation_level;time_series_type;calculation_id;calculation_type;calculation_execution_time_start;from_grid_area;calculation_result_id;metering_point_id
+804;8100000000115;;3.75;['calculated'];2023-02-01 12:00:00;total_ga;positive_grid_loss;;BalanceFixing;;;;571313180400100657
+804;8100000000115;;3.75;['calculated'];2023-02-01 12:15:00;total_ga;positive_grid_loss;;BalanceFixing;;;;571313180400100657
+804;8100000000115;;3.75;['calculated'];2023-02-01 12:30:00;total_ga;positive_grid_loss;;BalanceFixing;;;;571313180400100657
+804;8100000000115;;3.75;['calculated'];2023-02-01 12:45:00;total_ga;positive_grid_loss;;BalanceFixing;;;;571313180400100657
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/test_data/expected_results.csv:5]: Are we missing the balance responsible? In another PR? Should we add a task to the story?

Comment (BjarkeMeier) [reply source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/test_data/expected_results.csv:5]: I think quantity should be 3.750.

File: source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/states/scenario_state.py (hunk 6 above)
Comment (energinet-ajw) [reply source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/states/scenario_state.py]: New task.

File: source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/test_data/expected_results.csv (hunk 7 above)
Comment (energinet-ajw) [reply source/databricks/calculation_engine/tests/calculation_logic/features/energy_calculations/positive_grid_loss/when_period_is_one_hour/test_data/expected_results.csv:5]: New task.

**Review by BjarkeMeier — APPROVED** [review]
